shell/browser/api/electron_api_web_frame_main.cc

```diff
@@ -316,7 +316,7 @@ std::vector<content::RenderFrameHost*> WebFrameMain::FramesInSubtree() const {
 }
 
 void WebFrameMain::Connect() {
-  if (pending_receiver_) {
+  if (pending_receiver_ && render_frame_->IsRenderFrameCreated()) {
```

Comment:
I have questions..! Will this _reconnect_ later once the render frame is actually created? Will we miss any events or anything by not connecting immediately? Basically: is this the _correct_ fix?

Reply:
When I initially implemented this change, I think I may have overlooked the case where a speculative frame becomes active and we're sending messages using the `frame.send()` API. I do wonder if it's possible that messages aren't received when using `ipcRenderer.send` early in a preload script. I'd like to try creating a test to see whether we can break it this way.

[`WebContentsObserver::RenderFrameHostStateChanged(new_frame, old_state, new_state)`](https://source.chromium.org/chromium/chromium/src/+/main:content/public/browser/web_contents_observer.h;l=166-197;drc=cfb4b86c4d5138964cbcac7424ba3c4563775b49) is called when the speculative frame, or any other frame, changes to active.
```
  // When committing a cross-document, cross-RenderFrameHost navigation,
  // navigation-related callbacks are dispatched in the following order:
  // - RenderFrameHostStateChanged(new_frame, old_state, new_state)
  // - RenderFrameHostChanged(old_frame, new_frame)
  // - RenderFrameHostStateChanged(old_frame, old_state, new_state)
  // - DidFinishNavigation(navigation_handle)
```
If we do end up missing reconnecting the frame, listening for `new_state == contents::RenderFrameHost::LifecycleState::kActive` should help resolve it.

I'll followup once I've been able to experiment writing a test against early preload messages.

Reply:
I attempted several different methods of creating a WebContents (window.open, new BrowserWindow, window dispositions) and was unable to ever reproduce the crash. The stack trace included with the reported issue simply called `webContents.loadURL(url)` with a cross-origin URL. There may possibly be a race condition somewhere in Chromium leading to this behavior...

As for "is this the correct fix?", you might be right about the call to `GetInterface(pending_receiver_)` never happening. To ensure that it eventually gets called, I made it so `MaybeSetupMojoConnection()` will always check this when we need to get the renderer API. 4a91a92